### PR TITLE
162413776 extendedjob starting jobs

### PR DESCRIPTION
**Pull request by manno**
Supports pod template specs to define own pod. Add owner to created jobs to profit from garbage collection.

* fix event list query by adding involvedObject.kind to cache index
* extended jobs can specify what pod to run as a job
* jobs created are now labeled with 'extendedjob=true'
* fix bugs, i.e. pods with multiple labels
* improve logging
* more documentation

[#162413776](https://www.pivotaltracker.com/story/show/162413776)

### CODE REVIEW

File: pkg/kube/controllers/extendedjob/runner.go [hunk 1]
```diff
@@ -72,19 +88,37 @@ func (r *RunnerImpl) Run() {
...
+		// However we update the timestamp in the first iteration of
+		// this loop. Further iterations won't run the same job twice
+		// on the same pod, since the updated timestamp will be newer
+		// than any event we look at in this loop.
 		for _, podEvent := range filtered {
-			if r.isFreshStamp(extJob, podEvent) {
+			// this might be too noisy
+			r.log.Debugf("%s: looking at pod event: %s/%s for pod %s", extJob.Name, podEvent.Event.Name, podEvent.Event.Reason, podEvent.podName())
+
+			if podEvent.shouldTrigger(extJob.Name) {
```
Comment (aduffeck) [inline pkg/kube/controllers/extendedjob/runner.go]: I'm confused why this is not `if !podEvent.shouldTrigger(extJob.Name)`. Is this actually correct?

Comment (manno) [reply pkg/kube/controllers/extendedjob/runner.go]: yes

File: pkg/kube/controllers/extendedjob/pod_event.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+		podStamp := time.Unix(n, 0)
+		eventStamp := pe.Event.LastTimestamp.Time
+		return eventStamp.Before(podStamp)
+	}
+	return false
+}
+
+// updateTimestamp sets the timestamp on a pod to the current time
+func (pe PodEvent) updateTimestamp(client client.Client, exJobName string) error {
```
[the remark this answers is not in the document]
Comment (manno) [reply pkg/kube/controllers/extendedjob/pod_event.go]: yes

File: integration/environment/machine.go [hunk 3]
```diff
@@ -357,6 +359,72 @@ func (m *Machine) PodLabeled(namespace string, name string, desiredLabel, desire
...
+			return false, errors.Wrapf(err, "failed to query for jobs by label: %s", labels)
+		}
+
+		for _, job := range jobs.Items {
+			found[job.GetName()] = job
+		}
+		return len(found) == n, nil
+	})
+
+	if err == nil && len(found) == n {
```
[the remark this answers is not in the document]
Comment (manno) [reply integration/environment/machine.go]: No, too many jobs indicate a test failure.
But maybe I should check that in the test and not here.

**Pull request merged**